### PR TITLE
Feat: support more expressions in if and op.Fail

### LINKED ISSUE #4087: [Feature] Support more expressions in if
**Is your feature request related to a problem? Please describe.**

Right now we support `if` in the workflow step, we can support more expressions in if like:

```yaml
workflow:
  steps:
      - name: A
         type: A
      - name: B
         type: B
         if: A.status.phase == failed && A.status.reason == timeout
```

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### LINKED ISSUE #4088: [Feature] Support cue actions like `op.Failed`
**Is your feature request related to a problem? Please describe.**

In some cases, we may want this step to fail if the fields in the resource are not what we want.

That's when we need a new cue action like `op.Failed`, we can use this to set the steps' phase to fail to control the workflow.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by FogDong**
### Description of your changes

Fixes #4087 
Fixes #4088 

I have:

- [ ] Read and followed KubeVela's [contribution process](https://github.com/kubevela/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/kubevela/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: pkg/workflow/tasks/custom/task.go [hunk 1]
```diff
@@ -236,40 +219,133 @@ func (t *TaskLoader) makeTaskGenerator(templ string) (wfTypes.TaskGenerator, err
...
 		}
 		contextTempl = fmt.Sprintf("\ncontext: {%s}\ncontext: stepSessionID: \"%s\"", ms, id)
 	}
+	if pCtx == nil {
+		return ""
+	}
 	c, err := pCtx.ExtendedContextFile()
 	if err != nil {
-		return nil, err
+		return ""
```
Comment (leejanee) [inline pkg/workflow/tasks/custom/task.go:328]: pCtx.ExtendedContextFile() 返回的模版是以`context: `开头的吧,如何避免和workflow context生成的模版(也是 `context`开头)，统一渲染时因为path导致的冲突

Comment (FogDong) [reply pkg/workflow/tasks/custom/task.go:334]: workflow context 里面目前只有 application 的部分 metadata，pCtx.ExtendedContextFile() 也是我们内置的 context，目前不会冲突，之前的逻辑也是都用 context 开头，要改动的话影响比较大。

File: pkg/workflow/hooks/data_passing.go [hunk 2]
```diff
@@ -43,25 +43,15 @@ func Input(ctx wfContext.Context, paramValue *value.Value, step v1beta1.Workflow
...
-
+func Output(ctx wfContext.Context, taskValue *value.Value, step v1beta1.WorkflowStep, status common.StepStatus) error {
+	if wfTypes.IsStepFinish(status.Phase, status.Reason) {
 		for _, output := range step.Outputs {
 			v, err := taskValue.LookupByScript(output.ValueFrom)
 			if err != nil {
-				return err
+				v, err = taskValue.MakeValue("null")
+				if err != nil {
+					return err
```
[the remark this answers is not in the document]
Comment (FogDong) [reply pkg/workflow/hooks/data_passing.go:53]: 怎么样算表达格式错误？

[the remark this answers is not in the document]
Comment (FogDong) [reply pkg/workflow/hooks/data_passing.go:56]: Fixed

File: apis/core.oam.dev/common/types.go [hunk 3]
```diff
@@ -216,19 +216,19 @@ type WorkflowState string
...
 	WorkflowStateSucceeded WorkflowState = "Succeeded"
 	// WorkflowStateFinished means workflow is end.
-	WorkflowStateFinished WorkflowState = "finished"
+	WorkflowStateFinished WorkflowState = "Finished"
 	// WorkflowStateExecuting means workflow is still running or waiting some steps.
-	WorkflowStateExecuting WorkflowState = "executing"
+	WorkflowStateExecuting WorkflowState = "Executing"
 	// WorkflowStateSkipping means it will skip this reconcile and let next reconcile to handle it.
-	WorkflowStateSkipping WorkflowState = "skipping"
+	WorkflowStateSkipping WorkflowState = "Skipping"
```
Comment (wonderflow) [inline apis/core.oam.dev/common/types.go:231]: will these changes cause incompatible?

File: cmd/core/main.go [hunk 4]
```diff
@@ -143,7 +143,7 @@ func main() {
 	flag.IntVar(&resourcekeeper.MaxDispatchConcurrent, "max-dispatch-concurrent", 10, "Set the max dispatch concurrent number, default is 10")
 	flag.IntVar(&workflow.MaxWorkflowWaitBackoffTime, "max-workflow-wait-backoff-time", 60, "Set the max workflow wait backoff time, default is 60")
 	flag.IntVar(&workflow.MaxWorkflowFailedBackoffTime, "max-workflow-failed-backoff-time", 300, "Set the max workflow wait backoff time, default is 300")
-	flag.IntVar(&custom.MaxWorkflowStepErrorRetryTimes, "max-workflow-step-error-retry-times", 10, "Set the max workflow step error retry times, default is 10")
+	flag.IntVar(&wfTypes.MaxWorkflowStepErrorRetryTimes, "max-workflow-step-error-retry-times", 10, "Set the max workflow step error retry times, default is 10")
```
Comment (wonderflow) [inline cmd/core/main.go:146]: why don't merge it into workflow package just like above?

File: pkg/workflow/providers/mock/mock.go [hunk 5]
```diff
@@ -39,3 +39,9 @@ func (act *Action) Wait(message string) {
 	act.Phase = "Wait"
 	act.Message = message
 }
+
+// Fail ...
```
Comment (wonderflow) [inline pkg/workflow/providers/mock/mock.go:43]: finish it

File: pkg/workflow/providers/workspace/workspace.go [hunk 6]
```diff
@@ -161,6 +161,16 @@ func (h *provider) Break(ctx wfContext.Context, v *value.Value, act types.Action
 	return nil
 }
 
+// Fail let workflow fail.
```
Comment (wonderflow) [inline pkg/workflow/providers/workspace/workspace.go:164]: Explain what will happen after the workflow failed

File: pkg/workflow/tasks/custom/task.go [hunk 7]
```diff
@@ -236,40 +222,135 @@ func (t *TaskLoader) makeTaskGenerator(templ string) (wfTypes.TaskGenerator, err
...
+			return nil, err
+		}
+		if err := json.Unmarshal(bt, &params); err != nil {
+			return nil, err
+		}
+	}
+	return params, nil
+}
+
+func makeStatusValue(ctx wfContext.Context, step v1beta1.WorkflowStep, pd *packages.PackageDiscover, template string, stepStatus map[string]common.StepStatus, pCtx process.Context) (*value.Value, error) {
```
Comment (wonderflow) [inline pkg/workflow/tasks/custom/task.go:275]:
```suggestion
func buildValueForStatus(ctx wfContext.Context, step v1beta1.WorkflowStep, pd *packages.PackageDiscover, template string, stepStatus map[string]common.StepStatus, pCtx process.Context) (*value.Value, error) {
```

File: pkg/workflow/tasks/custom/task.go [hunk 8]
```diff
@@ -236,40 +222,135 @@ func (t *TaskLoader) makeTaskGenerator(templ string) (wfTypes.TaskGenerator, err
...
+			continue
+		}
+		statusTemplate += fmt.Sprintf("%s: %s\n", strings.ReplaceAll(name, "-", "_"), status)
+	}
+	statusTemplate += contextTempl
+	statusTemplate += "\n" + inputsTempl
+	return value.NewValue(template+"\n"+statusTemplate, pd, statusTemplate)
+}
+
+func makeValue(ctx wfContext.Context, pd *packages.PackageDiscover, templ, id string, pCtx process.Context) (*value.Value, error) {
```
Comment (wonderflow) [inline pkg/workflow/tasks/custom/task.go:308]: name the function better:

* convertTemplate
* calculateTemplate

File: apis/core.oam.dev/common/types.go (hunk 3 above)
Comment (FogDong) [reply apis/core.oam.dev/common/types.go:231]: These states are only used for `switch state` and message in status.

**Review by wonderflow — APPROVED** [review]

**Review by leejanee — APPROVED** [review]

**Review by Somefive — APPROVED** [review]
LGTM

**Review by barnettZQG — APPROVED** [review]

**Pull request merged**